### ISSUE TITLE
Help needed: Change server port during runtime

### ISSUE THREAD

**Issue by cpvalente** [cpvalente/ontime#358]
Currently, we use port 4001 for the [ontime express server](https://github.com/cpvalente/ontime/blob/master/apps/server/src/app.ts#L137).
The express server is responsible for serving the react app bundle and hosting most of the APIs (REST, WebSocket)

The React app expects to be able to find these resources at that port.

The challenge here is that, were the user able to change the runtime port, we would need to communicate this to front-end during runtime. Something like:
- Using the UI, the user changes the port for ontime
- The nodejs server would restart using the new port for the express app
- The react client would then connect to the new port

If any of this fails, both the server and the react app need to fallback to the last working port

We are looking for suggestions on how to best achieve this behaviour. PRs on this are welcome

Refs #262

**Comment by cpvalente** [comment]
Hi @asharonbaltazar, first of all thank you for the initiative.
This is something that we really want as part of the application but I do not have a clear picture on what it would entail.


### Adding some more pointers to the description of the issue
We currently piggyback from the express app to serve our REST API and Websocket.
Ontime uses REST to distribute its persisted data (project file) and WebSocket to distribute the runtime data. These changes in WebSockets happen often and are initialised by the server

The [express server is locked to port 4001](https://github.com/cpvalente/ontime/blob/master/apps/server/src/app.ts#L134).
Changing the port here should be trivial, we also have [a field in the project file](https://github.com/cpvalente/ontime/blob/master/demo-db/db.json#L41) and [a form in the UI](https://github.com/cpvalente/ontime/blob/master/apps/client/src/features/modals/settings-modal/AppSettings.tsx#L68) ready for letting the user defining the port. We would also need to [redefine our constants](https://github.com/cpvalente/ontime/blob/master/apps/client/src/common/api/apiConstants.ts#L1)

### The (maybe) challenge
The UI needs to know which port to reach to get the react app, since the built static files are served by express.

It potentially is just a matter of the user changing the port in the URL like `localhost:4001` -> `localhost:9999`

**Some (maybe) edge cases**
- Most users consume the [app through electron](https://github.com/cpvalente/ontime/blob/master/apps/electron/electron.config.js#L7). we would need to know in electron which port is being used at start
- We would also need to make sure the [docker distribution](https://github.com/cpvalente/ontime/blob/master/Dockerfile#L20) remains working

In short, I am unsure on what challenges this feature actually poses. There are several moving targets to look after
It might well "just work"

Hope this helps

**Comment by asharonbaltazar** [comment]
BTW, I sent you an email from `xxxb@outlook.com`. Maybe this effort is worth investing somewhere else.

**Comment by asharonbaltazar** [comment]
Hey, not familiar with backend as well as the frontend, which is why I chose this issue. I'm gonna need some help, so hopefully, you don't mind!

This is my logical process:

1. `app.ts` has a function called `startServer`. Since the function is called after `initAssets`, it should have access to the database. So this works: 
```ts
const { serverPort } = DataProvider.getSettings();
const returnMessage = `Ontime is listening on port ${serverPort}`;

expressServer = http.createServer(app);

socket.init(expressServer);
```
2. Now, I would assume that once a user has sent in a new server port in the `postSettings` route, we should compare whether the server port has been changed:
```ts
const { serverPort } = DataProvider.getSettings();

if (serverPort !== req.body?.serverPort) {
 // logic to restart the server
}
``` 

This is where I'm stuck, and Google hasn't been helpful. Some are saying to run external libraries responsible for managing processes and restart the server when it's shut down, etc, but that sounds smelly. 
Any advice?

**Comment by cpvalente** [comment]
Hey @asharonbaltazar, apologies for the late reply. The notification got mixed up in between a bunch of other things I had to tend to in GitHub

I absolutely do not mind helping, you are also welcome to contact by email if you would like timely replies. We can further arrange some slack or discord chat if that would be beneficial.

When it comes to your question:

I believe the challenge is keeping all the parties in sync (electron / docker, UI and server). A change in port needs to propagate correctly and we do not have a single communication channel for this.
The restarting of the process itself could also be tricky (as you have outlined). I usually see this done with a process manager which we cannot use in the case of electron

I would suggest separating this task in two steps so that we can have an MVP soned and iterate:
1. allow user to change port (will need a restart to take effect)
2. allow user to change port at runtime

The first task should add the whole infrastructure for syncing the different parts of the application.

We can then continue looking at the challenge of restarting the app during runtime, I would suspect that we could create a service to contain the express app and manage this part

What do you think? Are you comfortable with the change of scope?

**Comment by asharonbaltazar** [comment]
Hm, seems a bit more difficult than I anticipated. I'll unassign myself, so I can continue contributing elsewhere.
Separating it into two issues makes sense and explicitly fleshes out the scope of what's required.

**Comment by cpvalente** [comment]
No problem @asharonbaltazar 
This is indeed a task with several moving parts.

What do you think about https://github.com/cpvalente/ontime/issues/431

It should get your feet wet with backend and is in the immediate roadmap

Otherwise https://github.com/cpvalente/ontime/issues/403 should be pretty straightforward

**Comment by asharonbaltazar** [comment]
Got it. I'll take 403 then!

**Comment by pSnehanshu** [comment]
I am new to this project and found it in r/node, so pardon my naivety. I am asking the obvious question as an outsider. Why would the user change the port at runtime and why would the project want to support this usecase?

**Comment by cpvalente** [comment]
Hi @pSnehanshu, thank you for your interest.

The app's ability to integrate with other software and hardware is key to its success.
Unfortunately we are not the only ones that used the node - express demo code as a start and got the port 4001 locked as our server port 😅 

As I mentioned in one of the above comments: 
It is key being able to change port
It is nice to do it at runtime

Changing the port on its own has all the challenges I was unable to architect. The app consists of a node server, react frontend and electron wrapper. If the server were to change the port, we would need to find a way to communicate this to the other parts of the app.
See some more context in [the comment above](https://github.com/cpvalente/ontime/issues/358#issuecomment-1577420028)

**Comment by pSnehanshu** [comment]
I read the previous comments and now I have some context. I have a few questions:

1. Is this a offline app? I see there are download options on the website. If I download it, and install it on my computer, can I use it without internet connection?
2. You said the software is mostly used as an Electron app. Yet you provide a Docker image. When running it via Docker, will the interface be accessible via web browser? Can the Electron app be used in this context and does the Docker container work without internet connection?
3. Is it supposed to be used as a local app or as a cloud based app by the users? I mean do you expect your users to install it on their own cloud server and make it accessible to multiple end users over the network? And are those end users expected to use it via web browser or the Electron app?
4. When accessing via web browser, what are the URLs of the frontend and the backend? I assume they both are on port 4001?

**Comment by cpvalente** [comment]
Hi @pSnehanshu. I hope I manage to reply with enough context. Please feel free to follow up if needed

> Is this a offline app? I see there are download options on the website. If I download it, and install it on my computer, can I use it without internet connection?

Yes, this is an offline app. The app is really a nodejs service which exposes a websocket and a react frontend in the network interfaces of the machine it runs in

> You said the software is mostly used as an Electron app. Yet you provide a Docker image. When running it via Docker, will the interface be accessible via web browser? Can the Electron app be used in this context and does the Docker container work without internet connection?

Following up on the previous context, when we build the app, we build the react frontend and distribute it with the node app. The node app then runs an express server which serves the frontend along with the rest API and websockets.
To improve UX and facilitate distribution we package this in electron. Electron starts the node service and presents the react app in its window. From the browser you can reach to the node service on port 4001 and get served a copy of the app

Alternatively, you can use docker and run the node app. The react UI is then served from there as usual and - as you mentioned - accessible from any web browser.

Is it supposed to be used as a local app or as a cloud based app by the users? I mean do you expect your users to install it on their own cloud server and make it accessible to multiple end users over the network? And are those end users expected to use it via web browser or the Electron app?

We know that some users host the app in their own servers. I imagine that people have different approaches here, but I believe the docker workflow would be the best. Meaning the users would then be able to reach their app instance from the web browser.
It is important noticing that, in the cases the app is distributed from docker, you can easily expose the container in any port you want, so this flow is not impactful. We just need to make sure not to break it 😄 

When accessing via web browser, what are the URLs of the frontend and the backend? I assume they both are on port 4001?

That is exactly the case.
When you reach the app at `<ip-address>:4001` you are reaching the express-node service. At the root, the [service will return the frontend](https://github.com/cpvalente/ontime/blob/master/apps/server/src/app.ts#L67).
The express service also [serves a rest API ](https://github.com/cpvalente/ontime/blob/master/apps/server/src/app.ts#L57)along with the [websocket server](https://github.com/cpvalente/ontime/blob/master/apps/server/src/adapters/WebsocketAdapter.ts#L45)

**Comment by pSnehanshu** [comment]
Thanks for your response.

Since the react app and the API is served from the same port, changing the port to say 8000 and then opening localhost:8000 in the browser should be fine, because it can work with relative addressing (I guess it already works because you can already set a custom port when running a docker container).

I have more questions:
* How and where does the value of the `serverPort` field defined in `AppSettings.tsx` get's saved?
* Can the electron initialization code access this value at the time of booting? If so, we can dynamically set [this](https://github.com/cpvalente/ontime/blob/master/apps/electron/main.js#L160).

**Comment by cpvalente** [comment]
Hi, great line of questioning, it seems to me that you asking the right questions and might just be able to find the right solution
The docker part seems fine because you can expose the container port in a different port externally. In the container all services are still using port 4001. I am not particularly strong in this field and am unsure if we can anticipate any issues with docker by making this dynamic

> How and where does the value of the serverPort field defined in AppSettings.tsx get's saved?

The server port is [currently taken from the URL](https://github.com/cpvalente/ontime/blob/master/apps/client/src/common/api/apiConstants.ts#L15). This works as long as the express app and the websocket come from the same port

> Can the electron initialization code access this value at the time of booting? If so, we can dynamically set [this](https://github.com/cpvalente/ontime/blob/master/apps/electron/main.js#L160).

Initially Iplanned to do this from [the file we use as project config](https://github.com/cpvalente/ontime/blob/master/demo-db/db.json#L41), maybe we can revisit that idea?
- the backend reads the package json anyway to get the other parts of the file, it could easily take the server port as well
- the frontend takes the port from the URL as now
- the electron app could get the port returned from one of the functions in the server? potentially startServer could returns its status and port?

What do you think?

**Comment by pSnehanshu** [comment]
> Hi, great line of questioning, it seems to me that you asking the right questions and might just be able to find the right solution

Thanks, I hope I will be able to find the solution.

> The server port is currently taken from the URL. This works as long as the express app and the websocket come from the same port

Actually I was asking if the user changes the port from that screen (AppSettings.tsx), where would that be stored. I found the answer by looking around the codebase. Looks like it is stored in a JSON file at some location depending on the OS, and you're using LowDB to interact with it.

With that knowledge, I am sure we can use LowDB to fetch this saved port from the JSON database, and use it to form the URL and pass it to the [loadURL](https://github.com/cpvalente/ontime/blob/db1155ce1a2778866506eecdbe4dae192ab35159/apps/electron/main.js#L163) function.

I think the port handling the frontend isn't a big issue. It can take it directly from `window.location`. And the backend can fetch this value from the JSON database using LowDB and bind to this port.

The bigger thing is how to tell Electron to load the proper port. Which I think we can do using LowDB.

> the electron app could get the port returned from one of the functions in the server? potentially startServer could returns its status and port?

Yes, this approach is good too. In this case, the `startServer` function will fetch the port value saved in JSON database using LowDB, start the server on this port, and return it so that it can be passed to `loadURL`.

What do you think?

**Comment by cpvalente** [comment]
> Actually I was asking if the user changes the port from that screen (AppSettings.tsx), where would that be stored.

a-ha, apologies for the misunderstanding.
You have got the gist of it.
The code in `ontimeController.ts` currently ignores that value. We would need to add it here and to `validateSettings` in `ontimeController.validate.ts`.
From here the code persists this data in the json through the `DataProvider ` (lowdb adapter)

> With that knowledge, I am sure we can use LowDB to fetch this saved port from the JSON database, and use it to form the URL and pass it to the [loadURL](https://github.com/cpvalente/ontime/blob/db1155ce1a2778866506eecdbe4dae192ab35159/apps/electron/main.js#L163) function.

Absolutely, this sounds very possible.
lowDB is just a wrapper around the json file. Since this is a one-time read operation we might prefer doing it directly with fs


> What do you think?

I think you might have figured out enough pieces of the puzzle to put this together and I thank you for this.
I would be inclined to prefer the second version (where we get the port from the function), only because it would avoid us adding a disk read operation. I have not weighed the options and don't have strong feelings either way.

Do you think you are able to make a proposal?

**Comment by pSnehanshu** [comment]
> a-ha, apologies for the misunderstanding.

No worries!

> The code in ontimeController.ts currently ignores that value. We would need to add it here

So you mean the port value isn't currently saved? It is ignored by the `postSettings` route?

> Since this is a one-time read operation we might prefer doing it directly with fs

I would prefer not to, since that would be too adhoc. Since there already exists a `DataProvider` class (which wraps lowdb), I will prefer using that. This will be consistent with how data fetching in other places in the codebase.

> I would be inclined to prefer the second version (where we get the port from the function), only because it would avoid us adding a disk read operation.

How will we avoid doing disk read with this method? The port value is saved in JSON database, so we must do a disk read. Or did I misunderstood?

> Do you think you are able to make a proposal?

Proposal as in?

**Comment by cpvalente** [comment]
> So you mean the port value isn't currently saved? It is ignored by the postSettings route?

Yes since we did not want the users changing this. But that should be an easy enough fix


From what I understood, we have to options to explore

1. the electron process looks at the json file and finds the port. it can move on from here
2. the electron process gets the port from the node server when it starts

For the first operation:
- react gets port from url
- express gets port from file 
- electron gets port from file 

For electron to be able to read the port we would need to be able to read this file from the user directory. This can be done with fs or with the DataProvider as you describe. 
It could be tricky to get the DataProvider to run in electron. I might be wrong

For the second option:
- react gets port from url
- express gets port from file
- electron gets port from express (returned from the `startServer `function)

What I meant was that in this option, we avoid electron needing to read the file at all (through fs or any other way)

It seems to me that, by now, you have a good understanding of the use case to see all the moving parts. So I attempted asking whether you would be interested in creating a PR. I am happy to help

**Comment by pSnehanshu** [comment]
> express gets port from file

But if ontime is running inside docker, we may not want to do this, and instead it should always run on 4001 inside the container. We may also hide this field from the UI so that it is not configurable. Do you know if it will be possible to know if ontime is running inside docker or not?

> electron gets port from express (returned from the startServer function)

Yeah exactly. I also prefer this option.

> whether you would be interested in creating a PR

Yeah, I will get on it.

**Comment by cpvalente** [comment]
> But if ontime is running inside docker, we may not want to do this, and instead it should always run on 4001 inside the container. We may also hide this field from the UI so that it is not configurable. Do you know if it will be possible to know if ontime is running inside docker or not?

We set an environment variable in docker, this should help for the backend, but would not be available for the browser (UI)
https://github.com/cpvalente/ontime/blob/master/Dockerfile#L5
https://github.com/cpvalente/ontime/blob/master/apps/server/src/setup.ts#L40

I don't have a clear mental picture of the docker edge cases. I imagine that if the user were to change the port, the docker configuration would need to change to match.

Since we know in the backend when we are inside docker, we could have the router refusing to change that and returning an error. It might be simpler than getting this into the UI

> Yeah, I will get on it.

Thank you! 🙏

**Comment by cpvalente** [comment]
Merged and tested in v2.7.0. Great work @pSnehanshu
